util/sync/rbmutex.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sync
+
+import (
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/tidb/util/fastrand"
+)
+
+const (
+	// number of reader slots; must be a power of two
+	rslots = 4096
+	// slow-down guard
+	nslowdown = 9
+)
+
+// A RBMutex is a reader biased reader/writer mutual exclusion lock.
+// The lock can be held by an many readers or a single writer.
+// The zero value for a RBMutex is an unlocked mutex.
+//
+// A RBMutex must not be copied after first use.
+//
+// RBMutex is based on the BRAVO (Biased Locking for Reader-Writer
+// Locks) algorithm: https://arxiv.org/pdf/1810.01553.pdf
+//
+// RBMutex is a specialized mutex for scenarios, such as caches,
+// where the vast majority of locks are acquired by readers and write
+// lock acquire attempts are infrequent. In such scenarios, RBMutex
+// performs better than the sync.RWMutex on large multicore machines.
+//
+// RBMutex extends sync.RWMutex internally and uses it as the "reader
+// bias disabled" fallback, so the same semantics apply. The only
+// noticeable difference is in reader tokens returned from the
+// RLock/RUnlock methods.
+type RBMutex struct {
+	readers      [rslots]int32
```

Comment:
This should be a global variable...

Reply:
BRAVO is not a shared lock. It is well known that Golang' RWMutex is consist of an atomic value and mutex. But BRAVO is consist of a list of atomic and mutex. it can improves reader-reader scalability.

Here is BRAVO‘s introduction.

```
Designers of modern reader-writer locks confront a difficult trade-off related to reader scalability. Locks that have a compact memory representation for active readers will typically suffer under high intensity read-dominated workloads when the "reader indicator" state is updated frequently by a diverse set of threads, causing cache invalidation and coherence traffic. Other designs use distributed reader indicators, one per NUMA node, per core or even per thread. This improves reader-reader scalability, but also increases the size of each lock instance and creates overhead for writers.

We propose a simple transformation, BRAVO, that augments any existing reader-writer lock, adding just two integer fields to the lock instance. Readers make their presence known to writers by hashing their thread's identity with the lock address, forming an index into a visible readers table and installing the lock address into the table. All locks and threads in an address space can share the same readers table. Crucially, readers of the same lock tend to write to different locations in the table, reducing coherence traffic. Therefore, BRAVO can augment a simple compact lock to provide scalable concurrent reading, but with only modest and constant increase in memory footprint.
```

Reply:
BTW, BRAVO is not very suitable for Golang. it has too much overhead when to generate token. it may be suitable with more CPU core servers like AMD 128c and have many read concurrency.

Reply:
> adding just two integer fields to the lock instance.

The design in paper adds `2*4` bytes to each lock instance, but this PR adds `4096*4` bytes.

Reply:
```
class BRAVOLock<T> :
    int RBias
    Time InhibitUntil
    T Underlying

## Shared global :
BRAVOLock * VisibleReaders [4096]
```